app/models/publication.rb

```diff
@@ -250,31 +250,23 @@ def authoritative_pmid_source?
 
   private
 
-    def add_all_identifiers_in_db_to_pub_hash
-      publication_identifiers.reload if persisted?
-      pub_hash[:identifier] = publication_identifiers.map(&:identifier)
-    end
-
-    def sync_identifiers_in_pub_hash_to_db
+    # doesn't actually write the Pub to DB, presumed to be part of before_save callback or explicit save
+    def sync_identifiers_in_pub_hash
       incoming_types = Array(pub_hash[:identifier]).map { |id| id[:type] }
       publication_identifiers.each do |id|
         next if id.identifier_type =~ /^legacy_cap_pub_id$/i # Do not delete legacy_cap_pub_id
-        id.delete unless incoming_types.include? id.identifier_type
+        publication_identifiers.delete(id) unless incoming_types.include?(id.identifier_type)
       end
 
       Array(pub_hash[:identifier]).each do |identifier|
         next if identifier[:type] =~ /^SULPubId$/i
-        i = publication_identifiers.find { |x| x.identifier_type == identifier[:type] } || PublicationIdentifier.new
-        i.assign_attributes certainty: 'confirmed',
-                            identifier_type: identifier[:type],
-                            identifier_value: identifier[:id],
-                            identifier_uri: identifier[:url]
-        if i.persisted?
-          i.save
-        else
-          publication_identifiers << i unless publication_identifiers.include? i
-        end
+        i = publication_identifiers.find { |x| x.identifier_type == identifier[:type] } # find includes not yet saved pub ids
+        i ||= publication_identifiers.find_or_initialize_by(identifier_type: identifier[:type])
+        i.certainty        = 'confirmed'
+        i.identifier_value = identifier[:id]
+        i.identifier_uri   = identifier[:url]
       end
+      pub_hash[:identifier] = publication_identifiers.map(&:identifier)
     end
```

Comment:
It seems like the `Array` iteration was previously modifying the `publication_identifiers` in the loop, but now it doesn't and nothing directly modifies the loop variable `identifier`.  If I read it correctly, now the iteration is a no-op.  I'm not yet clear on this, but this whole thing looks a bit like the `PublicationIdentifier.pubhash_update` method recently added.  Maybe it could be rewritten like:
```ruby
pub_hash[:identifier]).each do |identifier|
  # find or create the matching PublicationIdentifier record - say `pub_id`
  pub_id.pubhash_update
end
```
It might work better as a class method on `PublicationIdentifier` - something to be responsible for taking an pub_hash[:identifier] hash and doing the find, build or CRUD for it. Dunno.

Reply:
This line here is moved in consolidating one of the other `sync`-related methods, from here:
https://github.com/sul-dlss/sul_pub/pull/334/files/c41775a9a3e21751a4ae9775c6dc91e8c501ed78..8337b69867e95934b6476f15fcfa175368353a7a#diff-54a25b14a10ce4b168bc13af6acd3190L255

Reply:
- At ln 269, do we need a `publication_identifiers.reload` in there?
- I _think_ I get it; the loop is updating or creating PublicationIdentifier records so they match the pub_hash data; then after the loop all the pub_hash identifiers are replaced with the db-records content.